src/main/scala/chisel3/util/experimental/ForceNames.scala

```diff
@@ -0,0 +1,290 @@
+    * @param signal Signal to name
+    */
+  def apply(signal: => chisel3.Bits): Unit = {
+    annotate(new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = ForceNameAnnotation(signal.toTarget, signal.toTarget.ref)
+      override def transformClass: Class[_ <: Transform] = classOf[ForceNamesTransform]
+    })
+  }
+
+  /** Force the name of this instance to the name its given during Chisel compilation
+    * @param instance Instance to name
+    */
+  def apply(instance: chisel3.experimental.BaseModule, name: String): Unit = {
+    annotate(new ChiselAnnotation with RunFirrtlTransform  {
+      def toFirrtl = {
+        val t = instance.toAbsoluteTarget
+        ForceNameAnnotation(t, name)
+      }
+      override def transformClass: Class[_ <: Transform] = classOf[ForceNamesTransform]
+    })
+  }
+
+  /** Force the name of this instance to the name its given during Chisel compilation
+    * This will rename after potential renames from other Custom transforms during FIRRTL compilation
+    * @param instance Signal to name
+    */
+  def apply(instance: chisel3.experimental.BaseModule): Unit = {
+    annotate(new ChiselAnnotation with RunFirrtlTransform  {
+      def toFirrtl = {
+        val t = instance.toAbsoluteTarget
+        ForceNameAnnotation(t, instance.instanceName)
+      }
+      override def transformClass: Class[_ <: Transform] = classOf[ForceNamesTransform]
+    })
+  }
+}
+
+/** Links the user-specified name to force to, with the signal/instance in the FIRRTL design
+  * @throws CustomTransformException when the signal is renamed to >1 target
+  * @param target signal/instance to force the name
+  * @param name name to force it to be
+  */
+case class ForceNameAnnotation(target: IsMember, name: String)
+    extends SingleTargetAnnotation[IsMember] {
+  def duplicate(n: IsMember): ForceNameAnnotation = this.copy(target = n, name)
+
+  // Errors if renaming to multiple targets
+  override def update(renames: RenameMap): Seq[Annotation] = {
+    renames.get(target) match {
+      case None => List(this)
+      case Some(newTargets) if newTargets.size > 1 =>
+        throw CustomTransformException(
+          new FirrtlUserException(
+            s"Cannot force the name of $target to $name because it is renamed to $newTargets." +
+              " Perhaps $target is not a ground type?"
+          )
+        )
+      case Some(newTargets) => newTargets.map(t => duplicate(t))
+    }
+  }
+}
+
+/** Contains utility functions for [[ForceNamesTransform]] */
+object ForceNamesTransform {
+  /** Returns the [[IsModule]] which is referred to, or if a [[ReferenceTarget]], the enclosing [[IsModule]]
+    * @param a signal/instance/module
+    * @return referring IsModule
+    */
+  def referringIsModule(a: IsMember): IsModule = a match {
+    case b: ModuleTarget    => b
+    case b: InstanceTarget  => b.targetParent
+    case b: ReferenceTarget => b.copy(component = Nil).targetParent.asInstanceOf[IsModule]
+  }
+
```

Comment:
All of these utilities seem pretty generic, should we add them to FIRRTL? That shouldn't preclude using them here but seems like a more natural place. I think maybe making them private or package private so that we don't introduce a new public API while also introducing them to FIRRTL makes more sense.